### DISCUSSION TITLE
Algoritmo de selección

### DISCUSSION THREAD

**Discussion by leandroj1** [leandroj1/ProyectoFinalPOO#26] [Ideas]
# Detalles del algoritmo de selección
El algoritmo tiene que recibir dos argumentos: `SolicitudEmpresa solicitudEmpresa, float porcentajeMatchRequerido`, donde el porcentaje es un número tal que `porcentajeMatchRequerido >= 0.0f && porcentajeMatchRequerido <= 100.0f`.

## 1. ¿Cómo se busca el personal?
Se recorre el *array list* general de personas, y se evalúa en el "nivel más alto", es decir, se evalúa si el objeto `Personal persona` (cualquiera de las clases derivadas) actual cumple con los requisitos de la solicitud, tanto los generales como los específicos.

**Nota: Los requisitos de peso para esta parte y la próxima es que el personal sea del tipo requerido y tenga los campos específicos (como `areaTecnica`).**

En caso de no cumplir a nivel general, se evalúan las solicitudes de `persona` para saber si hay alguna que cumpla con los  requerimientos de la solicitud. Sin embargo, se añade un peso: que la solicitud esté **activa**.

## 2. ¿Cómo se acumula el porcentaje?
Se requiere una cantidad de requisitos `n` para saber qué porcentaje le toca a cada propiedad. Así pues, hay propiedades que son "por defecto", es decir, que siempre van. Estas son:
```java
private  String  tipoPersonalSolicitado;
private  String  sexo;
private  float  salarioMax;
private  float  salarioMin;
private  int  edad;
private  int  agnosExperiencia;
private  boolean  disponibilidadSalirCiudad;
private  boolean  disponibilidadCambioResidencia;
private  String  tipoDeTrabajo;
private  String  nacionalidad;
private  boolean  esCasado;
private  ArrayList<String> idiomas;
```
**Nota: si sexo es `N/A`,  se toma en cuenta como una propiedad a favor de las personas para ser seleccionadas.**

En total son 12 propiedades (`idiomas`, al igual que otros *array lists* como `oficios`, cuenta como 1 sola), por lo que el valor inicial de la cantidad de requisitos es esta. (Se requiere implementar un método para esta parte)
Ahora, vienen las propiedades específicas:
```java
private  String  carrera;
private  String  universidad;
private  String  areaTecnica;
private  ArrayList<String> oficios;
```

A pesar de que se puede sumar evaluando si cada una es `null`, para evitar errores se sumará de acuerdo al tipo de personal. En otras palabras, si el personal que estoy evaluando es de tipo `Universitario`, en la cantidad de requisitos solo voy a sumar `carrera` y `universidad`.
De forma general, se tendrá una cantidad de requisitos igual a `12 + k`, donde el valor de `k` corresponde a la suma de los requisitos específicos. 
**Nota: se sumarán los requisitos específicos de acuerdo a si son `null` o no, esto para evitar errores.**

Asimismo, cada vez que se cumpla una propiedad, se acumulará un porcentaje  igual a `1.0f / (12 + k)`. En el caso de los listados como `idiomas` y `oficios`, se sumará `(1.0f / (12 + k)) / sizeArrayList`, para mantener la proporción y que no pase de `100.0f`.

Al final, se retornará un porcentaje que se multiplicará por `100.0f`, en vista de que todo el porcentaje que se acumuló con propiedades da un número entre `0.0f` y `1.0f`.

## 3. ¿Cómo se evalúa cada propiedad?
Las que son de tipo `String` se comparan con `equalsIgnoreCase()`. En el caso de los de tipo `int` o `float`, se evalúa si los datos de la persona son mayores o iguales que los pedidos, esto en vista de que la mayoría de propiedades son del tipo "mínimo". Pero para los salarios, se evaluara si el salario esperado por la persona está dentro del rango salarial ofrecido por la empresa. Para los `boolean`, simplemente se compara el estado.

## 4. Proceso de selección
En el caso de una persona, si su porcentaje resultante es mayor igual al requerido por la empresa, se guardará en un *array list* que retornará el algoritmo.

Si esté evaluando una solicitud, se creará un objeto `Personal` para que se pueda agregar al *array list* (pero únicamente servirá para mostrar el listado de candidatos de la solicitud), dado que en caso de se seleccione, se contratará por la [cédula de la persona, el RNC de la empresa y el id de la solicitud del personal](https://github.com/leandroj1/ProyectoFinalPOO/blob/27ab8bc55eb04a1a6f1027eb8955e32161a2f050/ProyectoFinal/src/logico/BolsaTrabajo.java#L110).

Edit: son 12 propiedades por _default_, faltaba nacionalidad.

**Comment by Brigibel12** [comment]
Considero que está bien planteado, sin embargo, me gustaría confirmar que pasaría en está situación: Si como empresa quiero a una persona que hable: Inglés, francés, español y alemán, y la persona solo habla inglés y español en vez de sumarle 1 sería 0.5? Lo mismo para la parte de los oficios.

**Reply by leandroj1** [reply]
> Si como empresa quiero a una persona que hable: Inglés, francés, español y alemán, y la persona solo habla inglés y español en vez de sumarle 1 sería 0.5? Lo mismo para la parte de los oficios.

Por ejemplo, si se pide un `Tecnico`, son 12 propiedades que se toman en cuenta para el algoritmo. Por que si solo habla inglés y español, se sumaría `(1.0f / 12) / 4 + (1.0f / 12) / 4`, el porcentaje correspondiente a la cantidad de idiomas que habla.

**Reply by Brigibel12** [reply]
> Por ejemplo, si se pide un `Tecnico`, son 12 propiedades que se toman en cuenta para el algoritmo. Por que si solo habla inglés y español, se sumaría `(1.0f / 12) / 4 + (1.0f / 12) / 4`, el porcentaje correspondiente a la cantidad de idiomas que habla.

Gracias!

**Comment by Brigibel12** [comment]
Para el paso 1, en vez de buscar el personal en la lista general de personas, no debería ser en el ArrayList de solicitudes tomando en cuenta que esté activa?, ya que, como algunos valores van a cambiar dependiendo de la solicitud que realice la persona, por ejemplo la parte de la disponibilidad de salir de la ciudad, sería más factible evaluarlo por ese medio.

**Reply by leandroj1** [reply]
Mmmm, cierto. En vista de que se guarda una solicitud con los datos de `Personal` cuando se crea el objeto, se pude evaluar solo el _array list_ de [solicitudes de personal](https://github.com/leandroj1/ProyectoFinalPOO/blob/27ab8bc55eb04a1a6f1027eb8955e32161a2f050/ProyectoFinal/src/logico/BolsaTrabajo.java#L22). De hecho, sería muchísimo más fácil, gracias.

**Reply by leandroj1** [reply]
> Mmmm, cierto. En vista de que se guarda una solicitud con los datos de `Personal` cuando se crea el objeto, se pude evaluar solo el _array list_ de [solicitudes de personal](https://github.com/leandroj1/ProyectoFinalPOO/blob/27ab8bc55eb04a1a6f1027eb8955e32161a2f050/ProyectoFinal/src/logico/BolsaTrabajo.java#L22). De hecho, sería muchísimo más fácil, gracias.

Aunque, en este caso, lo mejor sería retornar un [Map](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html) con los datos de Personal y Solicitud, para así evitar crear nuevos objetos temporales. Es conveniente para mostrar las personas que cumplen con una solicitud. 

Además, `Personal` no guarda datos, como la modalidad del trabajo, que se necesitan para mostrar el listado.

**Reply by leandroj1** [reply]
Sería algo más o menos así:
```java
public Map<Personal, SolicitudPersonal> getPersonalByPorcentajeMatch(SolicitudEmpresa solicitudEmpresa, float porcentajeMatchRequerido) {
    Map<Personal, SolicitudPersonal> candidatos = new HashMap<Personal, SolicitudPersonal>();

    if (solicitudEmpresa != null && (porcentajeMatchRequerido >= 0.0f && porcentajeMatchRequerido <= 100.0f)) {
        this.personal.forEach(person -> {
            person.getSolicitudes().forEach(solicitud -> {
                float resultPorcentaje = getPorcentajeMatchFrom(person, solicitud, solicitudEmpresa);
                if (resultPorcentaje >= porcentajeMatchRequerido) {
                    // Asignar porcentaje de match para no calcularlo de nuevo
                    solicitud.setPorcentajeMatchAsignado(resultPorcentaje);

                    candidatos.put(person, solicitud);
                }
            });
        });
    }

    return candidatos;
}
```

Edit: el método `setPorcentajeMatchAsignado` es para guardar el porcentaje de match asociado a esa solicitud, eso es para no calcularlo de nuevo en la interfaz. Es algo que estoy probando a ver si va bien.
